Enable icount mode for failure-prone platforms

wentongwu opened the pull request:
changes:
1. add cpu time in sanity-check for qemu thread
2. Enable icount mode for mps2_an385, mps2_an521 and qemu_x86, icount shift
    value should be selectd based on cpu clock frequency of every platform.
    The virtual cpu will execute one instruction every 2^shift ns of virtual
    time.
3. remove CONFIG_QEMU_TICKLESS_WORKAROUND for some timer test cases.
4. In Qemu icount mode, busy wait will cause lots of wall time and it's
    very easy to get sanitycheck timeout(this case will be successful if
    given enough timeout value for sanitycheck), so reduce test interval
    to save execution time.

have done:
- already test this PR on Unbuntu18.04 with sanity-check running at the same time on 10 different folders, some of which apply the patch and others don't, and see a lot of failed cases in the not enable icount folders, see very very little timeout cases on icount enabled folders(but I'm sure they will success if given large timeout value in sanity-check script).

need address:
- [x] tests/kernel/mem_protect/stack_random/kernel.memory_protection.stack_random always fail on qemu_x86 platform. I take a look the file rand32_timestamp.c used by x86 platform, not familiar with rdtsc instruction(__asm__ volatile("rdtsc" : "=a" (rv) : : "%edx");), maybe I need take a look x86 doc and review qemu code to see how qemu handle this on x86 platform, appreciate any help.
- [x] 7 test cases fail on qemu_x86_64 platform with latest commit.
- [ ] find very little timeout cases when icount enabled, that's maybe because on very heavy machine the cpu time is not accurate, maybe we can give large timeout value in sanity-check. 
- [x] tests/kernel/context/kernel.common always fail on qemu_cortex_r5 platform.
- [x] tests/kernel/mem_slab/mslab_threadsafe/kernel.memory_slabs.threadsafe always timeout on qemu_cortex_m0
- [ ] only platform qemu_nios2 doesn't work well for icount mode.
- [x] will test this PR with same method on Fedora. 

Fixes #14173

galak commented:
Can you make this change to qemu_cortex_r5 as well, see another of hangs there as well.

andyross commented:
Nice.  Did this suddenly start working on recent qemu versions?  When struggling with this ~1.5 years back, it was routine that some tests would get into a state where interrupts would stop arriving.

andyross commented on tests/kernel/common/prj.conf:
```diff
@@ -2,7 +2,6 @@ CONFIG_ZTEST=y
 CONFIG_PRINTK=y
 CONFIG_LOG=y
 CONFIG_POLL=y
-CONFIG_QEMU_TICKLESS_WORKAROUND=y
```
Does this remove this workaround for all tests that use it?  If not, it should, and we should remove the whole feature.  It only exists because of timer nondeterminism.

If not, and some things still seem to need it, it implies we have some more work to do somewhere.

wentongwu replied:
@andyross In this PR, I remove CONFIG_QEMU_TICKLESS_WORKAROUND in all tests cases, but just three Qemu platforms enabled icount so far, so maybe other Qemu platforms also need it.

andyross replied:
OK, then if it needs to stay we probably want to force it off at the kconfig level somehow so tests can say "give me the qemu tick workaround if the platform needs it" but the **platform**  layer somewhere can say "this configuration is running under qemu with -icount, don't hack the timer driver".

wentongwu replied:
Will change the patch, thanks

andrewboie commented:
Thanks for sending this, I will be running some tests today.

If this causes tests to take considerably more time to execute, we might want to consider tagging the time-sensitive tests, and only running those tests with icount, everything else with the current scheme.

wentongwu commented:
> Nice. Did this suddenly start working on recent qemu versions? When struggling with this ~1.5 years back, it was routine that some tests would get into a state where interrupts would stop arriving.

maybe there are some fix for that, started this work from last summer didn't find that kind of issue, but at the beginning I tried to enable every icount shift value(0-10) for qemu_x86 and mps2_an385, but most of that still have timing issue for Zephyr test cases, so have basic qemu code review(icount part) and gdb debug, maybe icount shift value should be selectd based on cpu clock frequency, both virtual cpu and real cpu(e.g. mps2_an385) should have most closet cpu frequency.

wentongwu commented:
> Thanks for sending this, I will be running some tests today.
> 
> If this causes tests to take considerably more time to execute, we might want to consider tagging the time-sensitive tests, and only running those tests with icount, everything else with the current scheme.

not every case will take more time to execute, but if Zephyr test case need wait some time interval(e.g. spin_for_ms), the execution time(wall time) will increase because of Qemu tcg mode and some extra code execution. 
And how is the test going on your side? @andrewboie

wentongwu commented:
> Can you make this change to qemu_cortex_r5 as well, see another of hangs there as well.

will try

andrewboie commented:
> tests/kernel/mem_protect/stack_random/kernel.memory_protection.stack_random always fail on qemu_x86 platform. I take a look the file rand32_timestamp.c used by x86 platform, not familiar with rdtsc instruction(asm volatile("rdtsc" : "=a" (rv) : : "%edx");), maybe I need take a look x86 doc and review qemu code to see how qemu handle this on x86 platform, appreciate any help.

rdtsc is just a CPU timestamp counter which places its result in EDX:EAX. https://c9x.me/x86/html/file_module_x86_id_278.html

It seems the problem here is that for whatever reason we're not getting any entropy. I'm taking a closer look today.

andyross commented:
Oooh, that reminds me: in fact one of the problems I was seeing when mucking with this long ago was that the timing sources were wrong in icount mode: the ratio between the HPET timer (used by the timer driver and k_cycle_get_32()) and the TSC was **different** depending on whether the app was running in icount or host timing mode.  So apps that touch the TSC on their own are likely to see different behavior.  In this particular case it would just be a different random seed, but it's something to check anyway.

I never did much to investigate; this wasn't as big a deal as the missing interrupts.

andrewboie commented:
We need a better way to get random numbers for QEMU tests.

Here's an assortment of "random" values produced by rand32_timestamp.c:

```
rand? bfe63a0
rand? bff84e0
rand? c00a680
rand? c01c8a0
rand? c02ea80
rand? c040c60
rand? c052e80
rand? c065060
rand? c077280
rand? c089460
rand? c09b640
rand? c0ad820
rand? c0bfa00
rand? c0d1c20
rand? c0e3e00
rand? c0f6020
rand? c108240
rand? c11a420
rand? c12c5c0
rand? c13e760
rand? c150900
rand? c162b20
rand? c174cc0
rand? c186e60
rand? c199000
rand? c1ab220
rand? c1bd3c0
rand? c1cf560
rand? c1e1700
rand? c1f38e0
rand? c205a80
rand? c217c60
rand? c229e00
rand? c23bfe0
rand? c24e180
rand? c260320
rand? c272500
rand? c2846e0
rand? c296880
rand? c2a8a20
rand? c2babc0
rand? c2ccd60
rand? c2def00
rand? c2f10e0
rand? c3032c0
rand? c3154a0
rand? c327640
rand? c3397e0
rand? c34b980
rand? c35db20
rand? c36fcc0
rand? c381e60
rand? c394000
rand? c3a6220
rand? c3b83c0
rand? c3ca560
rand? c3dc700
rand? c3ee8e0
rand? c400a80
rand? c412ca0
rand? c424e40
rand? c436fe0
rand? c449180
rand? c45b320
```

There is a pretty coarse granularity to TSC values reported, note that the low 4 bits are always zero and the lack of odd values in bits 5...8. This gets compounded by thread.c's adjust_stack_size() by using a modulo operation (in this case, 64). The computed fuzz values tend to be uniform. There's actually a bug open on this #6493

I'm wondering if it would be better for testing scenarios to use a xoroshiro generator that gets seeded with some predetermined value, at least we'd get a better distribution of values.

However, as a workaround, we can make this pass reliably if we increase CONFIG_STACK_POINTER_RANDOM=256 instead of the current 64 in prj.conf, we have a lot more bits to work with.

andrewboie commented:
With this trivial patch things are working reliably for me with qemu_x86:

```
diff --git a/tests/kernel/mem_protect/stack_random/prj.conf b/tests/kernel/mem_protect/stack_random/prj.conf
index ff20000218..98f14d8446 100644
--- a/tests/kernel/mem_protect/stack_random/prj.conf
+++ b/tests/kernel/mem_protect/stack_random/prj.conf
@@ -1,5 +1,5 @@
 CONFIG_ZTEST=y
-CONFIG_STACK_POINTER_RANDOM=64
+CONFIG_STACK_POINTER_RANDOM=256
 CONFIG_ENTROPY_GENERATOR=y
 CONFIG_TEST_RANDOM_GENERATOR=y
 
```

However, the build for qemu_x86_64 fails on nine tests. The result is thankfully reproducible, but the root cause(s) not yet clear:

```
INFO    - JOBS: 48
INFO    - Building initial testcase list...
INFO    - 167 test configurations selected, 578 configurations discarded due to filters.
INFO    - Adding tasks to the queue...
INFO    - Total complete:   42/ 167  25%  skipped:   10, failed:    0
ERROR   - qemu_x86_64               tests/kernel/mem_pool/mem_pool_threadsafe/kernel.memory_pool.threadsafe FAILED: Exited with 2
ERROR   - see: /dev/shm/sanity-out/qemu_x86_64/tests/kernel/mem_pool/mem_pool_threadsafe/kernel.memory_pool.threadsafe/handler.log
INFO    - Total complete:   43/ 167  25%  skipped:   10, failed:    1
ERROR   - qemu_x86_64               tests/kernel/mbox/mbox_usage/kernel.mailbox.usage  FAILED: Failed
ERROR   - see: /dev/shm/sanity-out/qemu_x86_64/tests/kernel/mbox/mbox_usage/kernel.mailbox.usage/handler.log
INFO    - Total complete:   48/ 167  28%  skipped:   11, failed:    2
ERROR   - qemu_x86_64               tests/kernel/interrupt/arch.interrupt              FAILED: Failed
ERROR   - see: /dev/shm/sanity-out/qemu_x86_64/tests/kernel/interrupt/arch.interrupt/handler.log
INFO    - Total complete:   62/ 167  37%  skipped:   11, failed:    3
ERROR   - qemu_x86_64               tests/kernel/lifo/lifo_usage/kernel.lifo.usage     FAILED: Failed
ERROR   - see: /dev/shm/sanity-out/qemu_x86_64/tests/kernel/lifo/lifo_usage/kernel.lifo.usage/handler.log
INFO    - Total complete:   80/ 167  47%  skipped:   13, failed:    4
ERROR   - qemu_x86_64               tests/kernel/mem_protect/userspace/kernel.memory_protection.userspace FAILED: Exited with 2
ERROR   - see: /dev/shm/sanity-out/qemu_x86_64/tests/kernel/mem_protect/userspace/kernel.memory_protection.userspace/handler.log
INFO    - Total complete:   99/ 167  59%  skipped:   17, failed:    5
ERROR   - qemu_x86_64               tests/kernel/spinlock/kernel.multiprocessing.spinlock FAILED: Timeout
ERROR   - see: /dev/shm/sanity-out/qemu_x86_64/tests/kernel/spinlock/kernel.multiprocessing.spinlock/handler.log
INFO    - Total complete:  100/ 167  59%  skipped:   17, failed:    6
ERROR   - qemu_x86_64               samples/portability/cmsis_rtos_v1/timer_synchronization/sample.portability.cmsis_rtos_v1.timer_synchronization FAILED: Timeout
ERROR   - see: /dev/shm/sanity-out/qemu_x86_64/samples/portability/cmsis_rtos_v1/timer_synchronization/sample.portability.cmsis_rtos_v1.timer_synchronization/handler.log
INFO    - Total complete:  101/ 167  60%  skipped:   17, failed:    7
ERROR   - qemu_x86_64               samples/portability/cmsis_rtos_v2/timer_synchronization/sample.portability.cmsis_rtos_v2.timer_synchronization FAILED: Timeout
ERROR   - see: /dev/shm/sanity-out/qemu_x86_64/samples/portability/cmsis_rtos_v2/timer_synchronization/sample.portability.cmsis_rtos_v2.timer_synchronization/handler.log
INFO    - Total complete:  110/ 167  65%  skipped:   18, failed:    8
ERROR   - qemu_x86_64               tests/kernel/smp/kernel.multiprocessing.smp        FAILED: Failed
ERROR   - see: /dev/shm/sanity-out/qemu_x86_64/tests/kernel/smp/kernel.multiprocessing.smp/handler.log
INFO    - Total complete:  167/ 167  100%  skipped:   18, failed:    9
INFO    - 140 of 149 tests passed (93.96%), 9 failed, 18 skipped with 0 warnings in 97.41 seconds
INFO    - In total 756 test cases were executed on 1 out of total 229 platforms (0.44%)
```

andrewboie commented:
If I switch to rand32_timestamp.c instead of rand32_timer.c I get what intuitively looks like a much better distribution:

```
rand? b2d13aa0
rand? 2a06ec6d
rand? a13c9e44
rand? 18725015
rand? 8fa801eb
rand? 6ddb3c2
rand? 7e13650f
rand? f54916e6
rand? 6c7ec8b7
rand? e3b47a87
rand? 5aea2c57
rand? d21fde28
rand? 49558ff8
rand? c08b41c9
rand? 37c0f39f
rand? aef6a576
rand? 262c5747
rand? 9d620917
rand? 1497baee
rand? 8bcd6cbe
rand? 3031e8f
rand? 7a38cfe2
rand? f16e81b3
rand? 68a43383
rand? dfd9e553
rand? 570f9724
rand? ce4548fb
rand? 457afacb
rand? bcb0ac9b
rand? 33e65e72
rand? ab1c1043
rand? 2251c219
rand? 998773ea
rand? 10bd25ba
rand? 87f2d791
rand? ff288961
rand? 765e3b32
```

I don't see why rand32_timer.c even exists. It's basically a carbon-copy of rand32_timestamp.c anyway. It works much worse. I'm going to simply delete it.

andrewboie commented:
I found some issues in tests/kernel/interrupt which I hopefully resolved here: #22958 

The qemu_x86_64 build with this PR passes if I set CONFIG_MP_NUM_CPUS=1, so whatever's causing the failures is at least tangentially SMP-related. I'm continuing to investigate.

wentongwu commented:
> I don't see why rand32_timestamp.c even exists. It's basically a carbon-copy of rand32_timer.c anyway. It works much worse. I'm going to simply delete it. #22951

Thanks for the fix!

stephanosio commented:
> * tests/kernel/context/kernel.common always fail on qemu_cortex_r5 platform.

```
===================================================================
starting test - test_kernel_cpu_idle

    Assertion failed at ../src/main.c:267: _test_kernel_cpu_idle: (tms2 < tms is true)
Bad ms per tick value computed,got 5528 which is less than 5529

FAIL - test_kernel_cpu_idle
===================================================================
```

@wentongwu This basically means that the `WFI` instruction in `k_cpu_idle` is not getting properly emulated (i.e. not causing processor halt until the next timer interrupt).

The following Xilinx QEMU source code lines show why that is:
https://github.com/Xilinx/qemu/blob/3b66f92d88a761803f24ff9bb3eb7b173832b2f4/target/arm/op_helper.c#L307-L312

The Xilinx QEMU previously ignored the `WFI` instruction completely due to co-simulation stability issues, which are irrelevant to the Zephyr CI testing, and I fixed that in the following commit:
https://github.com/zephyrproject-rtos/sdk-ng/commit/7c098e0d35edc86e0cfe22119dbfafe00ae58eb7

Unfortunately, in that fix, I left the `use_icount` check as is and that is why the `WFI` instruction is not halting the CPU.

stephanosio reviewed (commented):
Could you split https://github.com/zephyrproject-rtos/zephyr/pull/22904/commits/ac38b8f87a880746e12b096bf846208872f0f42a into separate commits? (one commit per board, preferably)?

wentongwu commented on boards/arm/qemu_cortex_r5/board.cmake:
```diff
@@ -9,6 +9,7 @@ set(QEMU_CPU_TYPE_${ARCH} cortex-r5)
 set(QEMU_FLAGS_${ARCH}
   -nographic
   -machine arm-generic-fdt
+  -icount shift=9,align=off,sleep=off -rtc clock=vm
```
@stephanosio thanks, I can't find the CPU info and ping you in previous github issue. So it's about 128Mhz?

stephanosio replied:
> So it's about 128Mhz?

Correct.

stephanosio replied:
With the `shift=3` and the https://github.com/zephyrproject-rtos/sdk-ng/pull/191 patch, I can see that sanitycheck runs very stable (ran 5 times, with no failures).

wentongwu replied:
@stephanosio  thanks for the confirm and fix.

wentongwu commented:
> Could you split [ac38b8f](https://github.com/zephyrproject-rtos/zephyr/commit/ac38b8f87a880746e12b096bf846208872f0f42a) into separate commits? (one commit per board, preferably)?

sure, will do that.

wentongwu commented:
> The following Xilinx QEMU source code lines show why that is:
> https://github.com/Xilinx/qemu/blob/3b66f92d88a761803f24ff9bb3eb7b173832b2f4/target/arm/op_helper.c#L307-L312
> 
> The Xilinx QEMU previously ignored the `WFI` instruction completely due to co-simulation stability issues, which are irrelevant to the Zephyr CI testing, and I fixed that in the following commit:
> [zephyrproject-rtos/sdk-ng@7c098e0](https://github.com/zephyrproject-rtos/sdk-ng/commit/7c098e0d35edc86e0cfe22119dbfafe00ae58eb7)
> 
> Unfortunately, in that fix, I left the `use_icount` check as is and that is why the `WFI` instruction is not halting the CPU.
> 

@stephanosio Thanks for the history and fix.

wentongwu committed:
tests: kernel: reduce test interval to save execution time

In Qemu icount mode, busy wait will cause lots of wall time and it's
very easy to get sanitycheck timeout(this case will be successful if
given enough timeout value for sanitycheck), so reduce test interval
to save execution time.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu commented:
two issues left: 1. CI takes very long time when icount enabled; 2. qemu_nios2 doesn't work well with icount, need more debug.

wentongwu committed:
tests: lib: heap: exclude qemu_riscv32 platform

This CPU-bound test on qemu_riscv32 platform is very slow when
QEMU icount mode enabled, taking upwards of several minutes.
There's little value here, this is a unit test of library code
and we have coverage of the RISC-V 32 bit arch via hifive1.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

stephanosio reviewed (commented):
Is there a reason why `qemu_cortex_a53` is not addressed in this PR? I see that there are quite a few failures for it (for the first run).

stephanosio reviewed (commented):
I see that there are quite a few tests that are on the verge of timing out (60s is the default timeout); some of them are related to this PR.

We should probably look into making the tests complete faster if feasible; if not, we should specify longer timeouts in the `testcase.yaml`s for these tests.

```
INFO    - 1364/1418 qemu_x86                  samples/userspace/prod_consumer/sample.userspace.prod_consumer PASSED (qemu 59.862s)

INFO    -  511/1418 qemu_cortex_r5            tests/kernel/sched/schedule_api/kernel.scheduler.multiq PASSED (qemu 59.963s)
INFO    -  690/1418 qemu_cortex_a53           tests/kernel/sched/schedule_api/kernel.scheduler   PASSED (qemu 59.952s)
INFO    -  691/1418 qemu_cortex_r5            tests/kernel/sched/schedule_api/kernel.scheduler   PASSED (qemu 59.993s)


INFO    -  227/1417 qemu_riscv32              tests/subsys/logging/log_immediate/logging.log_immediate PASSED (qemu 59.962s)
INFO    -  228/1417 qemu_riscv64              tests/subsys/logging/log_immediate/logging.log_immediate PASSED (qemu 59.890s)
INFO    - 1087/1417 qemu_xtensa               tests/lib/heap/lib.heap                            PASSED (qemu 59.868s)
INFO    - 1092/1417 qemu_x86_coverage         tests/lib/heap/lib.heap                            PASSED (qemu 59.883s)
INFO    - 1417/1417 qemu_riscv64              tests/subsys/logging/log_immediate/logging.log_immediate PASSED (qemu 59.961s)
```

wentongwu commented:
> Is there a reason why `qemu_cortex_a53` is not addressed in this PR? I see that there are quite a few failures for it (for the first run).

added qemu_cortex_a53, not intention.

wentongwu commented:
> We should probably look into making the tests complete faster if feasible;

agree, but I'm skeptical of loss accuracy to get speed. If some day we implement more timing cases which need higher accuracy, we may need re-configure it.

nashif commented on scripts/sanity_chk/sanitylib.py:
```diff
@@ -708,6 +708,15 @@ def __init__(self, instance, type_str):
 
         self.pid_fn = os.path.join(instance.build_dir, "qemu.pid")
 
+    @staticmethod
```
Please add a comment why you need that. you should consider using psutil hear instead of the custom mixed scripting.

wentongwu replied:
@nashif update

wentongwu replied:
@nashif CI environment doesn’t have psutil installed which caused shippable Timeout, is it possible setup psutil in CI in short time, though it’s easy on my PC?

andrewboie replied:
@wentongwu why haven't you added psutil to scripts/requirements.txt?

nashif replied:
@wentongwu  for now, just install it directly in shippable.yml, we will do the update soon.

wentongwu replied:
@nashif @andrewboie as indicated by https://github.com/giampaolo/psutil/blob/master/INSTALL.rst#linux, psutil depends on python3-dev on Ubuntu / Debian or python3-devel on RedHat / CentOS, which OS our CI is running?

nashif commented on scripts/sanity_chk/sanitylib.py:
```diff
@@ -737,9 +746,22 @@ def _thread(handler, timeout, outdir, logfile, fifo_fn, pid_fn, results, harness
         while True:
             this_timeout = int((timeout_time - time.time()) * 1000)
             if this_timeout < 0 or not p.poll(this_timeout):
+                if pid and this_timeout > 0:
```
Add comment about why this is needed here please.

wentongwu replied:
@nashif update

nashif commented on scripts/sanity_chk/sanitylib.py:
```diff
@@ -848,8 +867,29 @@ def handle(self):
+            try:
+                proc.wait(self.timeout)
+            except subprocess.TimeoutExpired:
+                #sometimes QEMU can't handle SIGTERM signal corrently
```
corrently? Did you mean correctly?

wentongwu replied:
@nashif yes, thanks, updated

andrewboie commented:
> > We should probably look into making the tests complete faster if feasible;
> 
> agree, but I'm skeptical of loss accuracy to get speed. If some day we implement more timing cases which need higher accuracy, we may need re-configure it.

@wentongwu I don't agree. I think some time needs to be spent on each of these long-running tests to determine whether the extended runtime is truly necessary.

Off the top of my head:

1. `samples/userspace/prod_consumer/sample.userspace.prod_consumer` I wrote this...there are lots of hard-coded delays which could be made much shorter. I will prepare a fix.
2. `tests/lib/heap/lib.heap` @andyross and @vonhust were talking about this in #24747 as it also takes a very long time to run and there might be opportunities to reduce the data size since it scales O(N^2)

I haven't looked at the scheduler or log_immediate tests yet, but someone should before we decide to simply kick up the timeout.

wentongwu committed:
scripts: sanitylib.py: enhancement for QEMUHandler

Sometimes QEMU can't handle SIGTERM signal correctly and just
run as normal, in that case kill -9 QEMU process directly and
leave sanitycheck judge the testing result by console output.

For failures caused by other reasons, terminate all releated
processes to avoid CI waits 2 hours and reports timeout.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
sanitycheck: Add cpu time for QEMUHandler

Add cpu time for QEMUHandler because the guest virtual time
in QEMU icount mode isn't host time and it's maintained by
counting guest instructions, we use QEMU process CPU time to
mostly simulate the time of guest OS.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu commented:
> > > We should probably look into making the tests complete faster if feasible;
> > 
> > 
> > agree, but I'm skeptical of loss accuracy to get speed. If some day we implement more timing cases which need higher accuracy, we may need re-configure it.
> 
> @wentongwu I don't agree. I think some time needs to be spent on each of these long-running tests to determine whether the extended runtime is truly necessary.
> 
@andrewboie my point here is that we should configure most approximate timer clock for every platform, say shift=5 is the most approximate one, just don't configure shift=8 to increase the execution speed. I agree `specify longer timeouts in the testcase.yamls for these long-running tests` or `some time needs to be spent on each of these long-running tests to determine whether the extended runtime is truly necessary`.

andrewboie commented:
OK. That does lead to a question from me: what criteria are we using to tune the shift values? and does that have any implications for the Kconfigs which specify the number of HW cycles per second?

wentongwu commented:
> OK. That does lead to a question from me: what criteria are we using to tune the shift values? and does that have any implications for the Kconfigs which specify the number of HW cycles per second?

yes, tune the freq of vCPU impacted by shift approximate 
 to SYS_CLOCK_HW_CYCLES_PER_SEC

stephanosio commented:
> > OK. That does lead to a question from me: what criteria are we using to tune the shift values? and does that have any implications for the Kconfigs which specify the number of HW cycles per second?
> 
> yes, tune the freq of vCPU impacted by shift approximate to SYS_CLOCK_HW_CYCLES_PER_SEC.

AFAIU, this is kind of like feeding a wrong clock frequency to a chip. As long as there is no on-chip physical timing constraint violation and external interactions that depend on certain timing requirements, it will work -- just slower or faster.

As for the external interaction timing requirements (which is not really all that important for the purpose of self-contained CI testing), we have already broken it by setting `sleep=off` (i.e. any `hlt`/`wfi` will cause QEMU to immediately jump to the next timeout, without actually waiting), so we might as well take the freedom of adjusting `shift` to whichever is the most advantageous.

andrewboie commented:
> As for the external interaction timing requirements (which is not really all that important for the purpose of self-contained CI testing), we have already broken it by setting `sleep=off` (i.e. any `hlt`/`wfi` will cause QEMU to immediately jump to the next timeout, without actually waiting), so we might as well take the freedom of adjusting `shift` to whichever is the most advantageous.

I concur. What's the limiting factor for larger shift values? After a certain point tests start to fail, is this because the timekeeping simply becomes too coarse?

wentongwu commented:
This only failed case is not related to this PR.

```
[127/180] Linking C static library zephyr/boards/arm/frdm_k64f/libboards__arm__frdm_k64f.a
[128/180] Building C object zephyr/drivers/adc/CMakeFiles/drivers__adc.dir/adc_shell.c.obj
FAILED: zephyr/drivers/adc/CMakeFiles/drivers__adc.dir/adc_shell.c.obj 
ccache /opt/sdk/zephyr-sdk-0.11.2/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc -DBOARD_FLASH_SIZE=CONFIG_FLASH_SIZE -DBUILD_VERSION=zephyr-v2.2.0-2438-gc0523967710b -DCPU_MK64FN1M0VLL12 -DKERNEL -DNDEBUG -D_FORTIFY_SOURCE=2 -D__PROGRAM_START -D__ZEPHYR_SUPERVISOR__ -D__ZEPHYR__=1 -I../../../../../../include -Izephyr/include/generated -I../../../../../../soc/arm/nxp_kinetis/k6x -I../../../../../../lib/fnmatch/. -I/home/buildslave/src/github.com/zephyrproject-rtos/modules/hal/cmsis/CMSIS/Core/Include -I/home/buildslave/src/github.com/zephyrproject-rtos/modules/hal/nxp/mcux/devices/MK64F12 -I/home/buildslave/src/github.com/zephyrproject-rtos/modules/hal/nxp/mcux/drivers/kinetis/. -isystem ../../../../../../lib/libc/minimal/include -isystem /opt/sdk/zephyr-sdk-0.11.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/include -isystem /opt/sdk/zephyr-sdk-0.11.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/9.2.0/include-fixed -Os -imacros/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/sanity-out/frdm_k64f/tests/boards/board_shell/boards.shell.shell/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -mcpu=cortex-m4 -mthumb -mabi=aapcs -imacros/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/include/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-address-of-packed-member -Wno-pointer-sign -Wpointer-arith -Wno-unused-but-set-variable -Werror=implicit-int -Werror -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/tests/boards/board_shell=CMAKE_SOURCE_DIR -fmacro-prefix-map=/home/buildslave/src/github.com/zephyrproject-rtos/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/home/buildslave/src/github.com/zephyrproject-rtos=WEST_TOPDIR -ffunction-sections -fdata-sections -std=c99 -nostdinc -MD -MT zephyr/drivers/adc/CMakeFiles/drivers__adc.dir/adc_shell.c.obj -MF zephyr/drivers/adc/CMakeFiles/drivers__adc.dir/adc_shell.c.obj.d -o zephyr/drivers/adc/CMakeFiles/drivers__adc.dir/adc_shell.c.obj   -c /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/adc/adc_shell.c
In file included from /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/adc/adc_shell.c:7:
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/adc/adc_shell.c:477:22: error: expected expression before ',' token
  477 |  ADC_SHELL_COMMAND(0),
      |                      ^
../../../../../../include/shell/shell.h:219:3: note: in definition of macro 'SHELL_STATIC_SUBCMD_SET_CREATE'
  219 |   __VA_ARGS__      \
      |   ^~~~~~~~~~~
[129/180] Linking C static library zephyr/lib/libc/minimal/liblib__libc__minimal.a
```

stephanosio commented:
> This only failed case is not related to this PR.

#25052

wentongwu commented:
@andrewboie networking cases will be excluded, new target patch will be updated soon.

wentongwu committed:
boards: qemu_x86: enable icount mode

Enable icount mode for qemu_x86 platform, The icount shift value is
selectd based on cpu clock frequency of this platform. The virtual
cpu will execute one instruction every 2^shift ns of virtual time.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
boards: qemu_cortex_m0: enable icount mode

Enable icount mode for qemu_cortex_m0 platform, The icount shift value
is selectd based on cpu clock frequency of this platform. The virtual
cpu will execute one instruction every 2^shift ns of virtual time.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
boards: qemu_cortex_m3: enable icount mode

Enable icount mode for qemu_cortex_m3 platform, The icount shift value
is selectd based on cpu clock frequency of this platform. The virtual
cpu will execute one instruction every 2^shift ns of virtual time.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
boards: qemu_cortex_a53: enable icount mode

Enable icount mode for qemu_cortex_a53 platform, The icount shift
value is selectd based on cpu clock frequency of this platform.
The virtual cpu will execute one instruction every 2^shift ns of
virtual time.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
boards: qemu_riscv32: enable icount mode

Enable icount mode for qemu_riscv32 platform, The icount shift value
is selectd based on cpu clock frequency of this platform. The virtual
cpu will execute one instruction every 2^shift ns of virtual time.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
boards: qemu_riscv64: enable icount mode

Enable icount mode for qemu_riscv64 platform, The icount shift value
is selectd based on cpu clock frequency of this platform. The virtual
cpu will execute one instruction every 2^shift ns of virtual time.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
boards: hifive1: enable icount mode

Enable icount mode for hifive1 platform, The icount shift value is
selectd based on cpu clock frequency of this platform. The virtual
cpu will execute one instruction every 2^shift ns of virtual time.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
boards: qemu_xtensa: enable icount mode

Enable icount mode for qemu_xtensa platform, The icount shift value
is selectd based on cpu clock frequency of this platform. The virtual
cpu will execute one instruction every 2^shift ns of virtual time.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu commented:
@andrewboie @stephanosio can you please review the updated patch? Thanks

andrewboie commented on .shippable.yml:
```diff
@@ -25,6 +25,7 @@ build:
         options: "-e HOME=/home/buildslave --privileged=true --tty --net=bridge --user buildslave"
 
     ci:
+      - sudo pip3 install psutil
```
why isn't psutil in scripts/requirements.txt?

wentongwu replied:
@andrewboie it has been added in another PR, I add here to make sure CI runs well and marked DNM

andrewboie replied:
I see it now, it's already in master. OK.

wentongwu replied:
@andrewboie but I think we can merge this PR right now.

nashif replied:
not yet, docker image needs to be updated first and this commit needs to be removed.

nashif commented:
@andrewboie  since the qemu_x86 change was reverted, is it safe to merge this PR as is?

wentongwu commented:
> @andrewboie since the qemu_x86 change was reverted, is it safe to merge this PR as is?

@nashif I think currently we should merge this, at least for kernel cases it works very well on the icount enabled qemu simulators, the only no good is we need more wall time to execute CI. For networking, I don't think we should enable icount because it only guarantees the deterministic execution from guest self point of view.

tyrantlizard77 reviewed (approved):
LGTM

As a future enhancement, I'd like the "shift" value to also be settable in Kconfig.

stephanosio commented on boards/arm/qemu_cortex_r5/board.cmake:
```diff
@@ -5,12 +5,15 @@
 set(QEMU_ARCH xilinx-aarch64)
 
+if(CONFIG_QEMU_ICOUNT)
+  set(ICOUNT_ARG -icount shift=3,align=off,sleep=off -rtc clock=vm)
+endif()
 set(QEMU_CPU_TYPE_${ARCH} cortex-r5)
 set(QEMU_FLAGS_${ARCH}
   -nographic
   -machine arm-generic-fdt
-  -icount shift=3,align=off,sleep=off -rtc clock=vm
   -dtb ${ZEPHYR_BASE}/boards/${ARCH}/${BOARD}/fdt-single_arch-zcu102-arm.dtb
+  ${ICOUNT_ARG}
   )
```
No need to give `qemu_cortex_r5` a special treatment, we should be able to just append here as well.

wentongwu replied:
@stephanosio thanks, updated.

wentongwu committed:
boards: remove the existing qemu icount configuration

Remove the existing qemu icount configuration because icount mode
will be controlled by Kconfig QEMU_ICOUNT so that none suitable
cases(especially networking cases) can exclude icount configuration

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu commented:
> As a future enhancement, I'd like the "shift" value to also be settable in Kconfig.

Thanks, sure, will do.

wentongwu committed:
timer: remove QEMU_TICKLESS_WORKAROUND

Qemu icount mode enabled, remove QEMU_TICKLESS_WORKAROUND.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

nashif commented on .shippable.yml:
```diff
@@ -25,6 +25,7 @@ build:
         options: "-e HOME=/home/buildslave --privileged=true --tty --net=bridge --user buildslave"
 
     ci:
+      - sudo pip3 install psutil
```
I think you should be able to remove this, it is already part of the docker..

wentongwu replied:
@nashif update

nashif commented on boards/Kconfig:
```diff
@@ -28,6 +28,15 @@ endchoice
 rsource "shields/*/Kconfig.shield"
 
 menu "Board Options"
+config QEMU_ICOUNT
```
the commit which has this change should be much earlier and before you start using QEMU_ICOUNT, for bisectability.

wentongwu replied:
@nashif update

nordic-krch commented:
Could it be that `qemu_arc_hs` and `qemu_arc_em` need the same fix? I see some tests failing on those platforms due to timing issues.

stephanosio commented:
> Could it be that `qemu_arc_hs` and `qemu_arc_em` need the same fix? I see some tests failing on those platforms due to timing issues.

@nordic-krch There is actually an issue for that: #26227.
